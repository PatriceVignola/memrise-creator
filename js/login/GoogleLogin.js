/* @flow */

import React from 'react';
import { WebView } from 'react-native';

import {
  type NavigationScreenProp,
  NavigationActions,
} from 'react-navigation';

import URL, { type Url } from 'url-parse';

const googleLoginUrl = 'https://www.memrise.com/accounts/login/google-oauth2/';
const webViewUserAgent = 'Mozilla/5.0 (compatible; Trident/6.0; ARM; Touch)';

type Props = {
  navigation: NavigationScreenProp<any>,
};

// TODO: Complete this def and move it into a libdef for WebView
export type StateChangeData = {
  url: string,
};

export default class GoogleLogin extends React.Component<Props> {
  redirectUrl: ?Url;

  constructor(props: Props) {
    super(props);

    this.onNavigationStateChange = this.onNavigationStateChange.bind(this);
  }

  onNavigationStateChange = (e: StateChangeData) => {
    const url = new URL(e.url, true);

    if (url.query.redirect_uri) {
      this.redirectUrl = new URL(url.query.redirect_uri);
    } else if (this.redirectUrl && url.pathname === this.redirectUrl.pathname) {
      this.props.navigation.dispatch(NavigationActions.reset({
        index: 0,
        actions: [NavigationActions.navigate({ routeName: 'CourseList' })],
      }));
    }
  }

  render() {
    return (
      <WebView
        source={{ uri: googleLoginUrl }}
        onNavigationStateChange={this.onNavigationStateChange}
        userAgent={webViewUserAgent}
      />
    );
  }
}
