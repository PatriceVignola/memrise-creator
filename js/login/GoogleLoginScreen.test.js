/* @flow */

import React from 'react';
import { addNavigationHelpers, NavigationActions } from 'react-navigation';
import { WebView } from 'react-native';
import Renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';

import GoogleLogin from './GoogleLogin';
import type { StateChangeData } from './GoogleLogin';

const shallowRenderer = new ShallowRenderer();

describe('GoogleLogin', () => {
  const mockDispatchFunction = jest.fn();

  const navigation = addNavigationHelpers({
    state: {
      key: 'key',
      routeName: 'routeName',
      path: 'path',
    },
    dispatch: mockDispatchFunction,
  });

  afterEach(() => {
    mockDispatchFunction.mockClear();
  });

  test('renders the Google Login', () => {
    const shallow = shallowRenderer.render((
      <GoogleLogin
        navigation={navigation}
      />
    ));
    expect(shallow).toMatchSnapshot();
  });

  test('navigates to CourseList after redirecting', () => {
    const mockResetObject = NavigationActions.reset({
      index: 0,
      actions: [NavigationActions.navigate({ routeName: 'CourseList' })],
    });

    const deep = Renderer.create((
      <GoogleLogin
        navigation={navigation}
      />
    ));

    const mockStateChangeUrl: StateChangeData = {
      url: 'https://url.com?redirect_uri=https://url.com/redirect_path',
    };

    const mockStateChangeRedirect: StateChangeData = {
      url: 'https://url.com/redirect_path',
    };

    deep.root.findByType(WebView).props.onNavigationStateChange(mockStateChangeUrl);
    deep.root.findByType(WebView).props.onNavigationStateChange(mockStateChangeRedirect);

    expect(mockDispatchFunction).toHaveBeenCalledTimes(1);
    expect(mockDispatchFunction).toHaveBeenCalledWith(mockResetObject);
  });

  test('does not redirect if the redirect url is not equal to redirect_uri', () => {
    const deep = Renderer.create((
      <GoogleLogin
        navigation={navigation}
      />
    ));

    const mockStateChangeUrl: StateChangeData = {
      url: 'https://url.com?redirect_uri=https://url.com/redirect_path',
    };

    const mockStateChangeBadRedirect: StateChangeData = {
      url: 'https://url.com/bad_redirect_path',
    };

    deep.root.findByType(WebView).props.onNavigationStateChange(mockStateChangeUrl);
    deep.root.findByType(WebView).props.onNavigationStateChange(mockStateChangeBadRedirect);

    expect(mockDispatchFunction).toHaveBeenCalledTimes(0);
  });
});
