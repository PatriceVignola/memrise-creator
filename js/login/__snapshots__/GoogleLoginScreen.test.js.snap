// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleLogin renders the Google Login 1`] = `
<WebView
  onNavigationStateChange={[Function]}
  scalesPageToFit={true}
  source={
    Object {
      "uri": "https://www.memrise.com/accounts/login/google-oauth2/",
    }
  }
  userAgent="Mozilla/5.0 (compatible; Trident/6.0; ARM; Touch)"
/>
`;
